---
title: "scATAC_analysis"
author: "xi luo"
date: "2023-11-28"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

```{r}
setRepositories(ind=1:3)
install.packages("Signac")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.18")
BiocManager::install("EnsDb.Hsapiens.v75")
install.packages('Seurat')
BiocManager::install("EnsDb.Mmusculus.v79")

```

```{r}
library(Signac)
library(Seurat)
library(EnsDb.Hsapiens.v75)
library(EnsDb.Mmusculus.v79)
library(ggplot2)
library(patchwork)
```

```{r}
counts <- Matrix::readMM("D:/Desktop/ATAC/matrix.mtx")
barcodes <- readLines("D:/Desktop/ATAC/barcodes.tsv")
peaks <- read.table("D:/Desktop/ATAC/peaks.bed", sep="\t")
peaknames <- paste(peaks$V1, peaks$V2, peaks$V3, sep="-")

colnames(counts) <- barcodes
rownames(counts) <- peaknames
```

```{r}
chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  fragments = 'D:/Desktop/ATAC/fragments/fragments.tsv.gz',
  genome = "mm10",
  min.cells = 1,
  min.features = 200
)

mouse_embryo <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks"
)

mouse_embryo[['peaks']]
granges(mouse_embryo)

```

```{r}
# extract gene annotations from EnsDb
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevels(annotations) <- paste0('chr', seqlevels(annotations))
genome(annotations) <- "mm10"
#seqlevelsStyle(annotations) <- "UCSC"
Annotation(mouse_embryo) <- annotations

# compute nucleosome signal score per cell
mouse_embryo <- NucleosomeSignal(object = mouse_embryo)

mouse_embryo$nucleosome_group <- ifelse(mouse_embryo$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
FragmentHistogram(object = mouse_embryo, group.by = 'nucleosome_group', region = 'chr1-1-10000000')

# compute TSS enrichment score per cell
mouse_embryo <- TSSEnrichment(object = mouse_embryo, fast = FALSE)

# add blacklist ratio and fraction of reads in peaks
mouse_embryo$pct_reads_in_peaks <- mouse_embryo$peak_region_fragments / mouse_embryo$passed_filters * 100
mouse_embryo$blacklist_ratio <- mouse_embryo$blacklist_region_fragments / mouse_embryo$peak_region_fragments
devtools::install_version(package = 'Signac', version = package_version('0.2.5'))

```

```{r}
DensityScatter(mouse_embryo, x = 'nCount_peaks', y = 'TSS.enrichment', log_x = TRUE, quantiles = TRUE)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
